<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to benchdamic - EuroBioC2022 workshop • benchdamicWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to benchdamic - EuroBioC2022 workshop">
<meta property="og:description" content="benchdamicWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">benchdamicWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction_to_benchdamic.html">An introduction to benchdamic - EuroBioC2022 workshop</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mcalgaro93/benchdamicWorkshop" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="introduction_to_benchdamic_files/kePrint-0.0.1/kePrint.js"></script><link href="introduction_to_benchdamic_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An introduction to benchdamic - EuroBioC2022
workshop</h1>
                        <h4 data-toc-skip class="author">Matteo
Calgaro<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mcalgaro93/benchdamicWorkshop/blob/HEAD/vignettes/introduction_to_benchdamic.Rmd" class="external-link"><code>vignettes/introduction_to_benchdamic.Rmd</code></a></small>
      <div class="hidden name"><code>introduction_to_benchdamic.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="an-introduction-to-benchdamic---eurobioc2022-workshop">An introduction to benchdamic - EuroBioC2022 workshop<a class="anchor" aria-label="anchor" href="#an-introduction-to-benchdamic---eurobioc2022-workshop"></a>
</h2>
<p>Author: Matteo Calgaro<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a></p>
<p>Last modified: 12 September, 2022.</p>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<div class="section level4">
<h4 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h4>
<p>This workshop provides an introductory example on how to work with
the analysis framework firstly proposed in <strong>“Assessment of
statistical methods from single cell, bulk RNA-seq, and metagenomics
applied to microbiome data”</strong> by <a href="https://doi.org/10.1186/s13059-020-02104-1" class="external-link">Calgaro et al.
(2020)</a>.</p>
<p>We will test a couple of methods for differential abundance (DA)
analysis on a microbiome dataset and we will see how to test custom
methods on the same dataset. Performances of each method are evaluated
with respect to i) suitability of distributional assumptions (GOF), ii)
ability to control false positives (TIEC), iii) concordance of the
findings, and iv) enrichment of DA microbial species in specific
conditions.</p>
<p>This workshop is structured as an instructor-led live demo but you
can decide whether to spend your time by following the examples as they
are or by playing with the large combinations of parameters offered by
each implemented DA method.</p>
<hr>
<blockquote>
<p>Moreover, during the workshop you will find indented sections like
this one. These are in-depth sections to explore after the workshop if
you are interested.</p>
</blockquote>
<hr>
</div>
<div class="section level4">
<h4 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h4>
<ul>
<li>Basic knowledge of R syntax;</li>
<li>Familiarity with the <code>phyloseq</code> class;</li>
<li>Familiarity with DA analysis in microbiome data.</li>
</ul>
<p>Some suggested background readings for the workshop:</p>
<ul>
<li>“Assessment of statistical methods from single cell, bulk RNA-seq,
and metagenomics applied to microbiome data” by <a href="https://doi.org/10.1186/s13059-020-02104-1" class="external-link">Calgaro et al.
(2020)</a>
</li>
<li>“Microbiome differential abundance methods produce different results
across 38 datasets” by <a href="https://doi.org/10.1038/s41467-022-28034-z" class="external-link">Nearing et al.
(2022)</a>;</li>
<li>“Normalization and microbial differential abundance strategies
depend upon data characteristics” by <a href="https://doi.org/10.1186/s40168-017-0237-y" class="external-link">Weiss et al.
(2017)</a>;</li>
</ul>
</div>
<div class="section level4">
<h4 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h4>
<p>The Bioconductor package we will use is named
<code>benchdamic</code>, acronym for “BENCHmarking of Differential
Abundance detection methods for MICrobial data”.</p>
</div>
<div class="section level4">
<h4 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h4>
<p>During the workshop we will have the first 35 minutes of active
coding (using the supplied examples). Questions are being addressed
during the final 10 minutes.</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Overview</td>
<td>2 - 3 mins</td>
</tr>
<tr class="even">
<td>GOF</td>
<td>5 mins</td>
</tr>
<tr class="odd">
<td>TIEC</td>
<td>12 mins</td>
</tr>
<tr class="even">
<td>Concordance</td>
<td>5 mins</td>
</tr>
<tr class="odd">
<td>Enrichment</td>
<td>9 mins</td>
</tr>
<tr class="even">
<td>Conclusions</td>
<td>1 min</td>
</tr>
<tr class="odd">
<td>Questions and Answer</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h4>
<p>During this workshop you will:</p>
<ul>
<li><p>see the main functions of <code>benchdamic</code>;</p></li>
<li><p>compare some DA methods on a simple microbiome dataset;</p></li>
<li><p>set up a custom method and include it into the
benchmarking;</p></li>
<li><p>evaluate DA methods’ performances using the graphical
outputs.</p></li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="workshop">Workshop<a class="anchor" aria-label="anchor" href="#workshop"></a>
</h3>
<div class="section level4">
<h4 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h4>
<p>Firstly we load some packages for basic functions and data:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">benchdamic</span><span class="op">)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'gplots':</span></span>
<span><span class="co">#&gt;   method         from     </span></span>
<span><span class="co">#&gt;   reorder.factor DescTools</span></span>
<span><span class="co"># Data management</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span></span>
<span><span class="co"># Graphics and tables</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/" class="external-link">kableExtra</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="data-loading">Data loading<a class="anchor" aria-label="anchor" href="#data-loading"></a>
</h4>
<p>During the workshop we will use the “ps_plaque_16S” dataset. In
contains 16S data from:</p>
<ul>
<li><p>30 participants of the Human Microbiome Project (HMP);</p></li>
<li><p>samples collected from subgingival plaque and supragingival
plaque for each subject (a total of 60 samples);</p></li>
<li><p>88 taxa, all features having the same genus-level taxonomic
classification have been collapsed together (a total of 88 taxa
corresponding to 88 genera).</p></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ps_plaque_16S"</span><span class="op">)</span></span>
<span><span class="va">ps_plaque_16S</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 88 taxa and 60 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 60 samples by 7 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 88 taxa by 6 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; phy_tree()    Phylogenetic Tree: [ 88 tips and 87 internal nodes ]</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="goodness-of-fit">Goodness of Fit<a class="anchor" aria-label="anchor" href="#goodness-of-fit"></a>
</h4>
<p><strong>Assumption</strong>: Many differential abundance detection
methods are based on parametric distributions.</p>
<p><strong>Question</strong>: Which are the parametric distributions
that can fit both the proportion of zeros and the counts in your
data?</p>
<hr>
<blockquote>
<p><strong>GOF structure</strong></p>
<p>Several distributions are available in <code>benchdamic</code>: 1)
the negative binomial (<strong>NB</strong>) distribution used in
packages like <code>edgeR</code> or <code>DESeq2</code>, 2) the
zero-inflated negative binomial (<strong>ZINB</strong>) distribution as
implemented by the <code>zinbwave</code> package, 3) the truncated
Gaussian Hurdle model as implemented by the <code>MAST</code> package 4)
the zero-inflated Gaussian (<strong>ZIG</strong>) mixture model as
implemented by the <code>metagenomeSeq</code> package, and 5) the
Dirichlet-Multinomial (<strong>DM</strong>) distribution which is the
multivariate extension of the beta-binomial (<em>e.g.</em> used by the
<code>corncob</code> package).</p>
<p>The relationships between the functions that can be used in this
section are explained in the next diagram. To help with the reading:
green boxes represent the inputs or the outputs, red boxes are the
methods and blue boxes are the main parameters of those method.</p>
<div class="figure">
<img src="GOF_structure.svg" alt=""><p class="caption">Goodness of Fit diagram</p>
</div>
</blockquote>
<hr>
<div class="section level5">
<h5 id="fit-parametric-distributions">Fit parametric distributions<a class="anchor" aria-label="anchor" href="#fit-parametric-distributions"></a>
</h5>
<p>We choose one of the two body sites in order to fit the models on a
homogeneous group of samples. The same analysis can be performed on both
body subsites, or in general, in each group of samples in your data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Subset Supragingival Plaque samples</span></span>
<span><span class="va">ps_sup</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">ps_plaque_16S</span>, <span class="va">HMP_BODY_SUBSITE</span> <span class="op">==</span> <span class="st">"Supragingival Plaque"</span><span class="op">)</span></span>
<span><span class="co"># Remove taxa with all zeroes</span></span>
<span><span class="va">ps_sup</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/filter_taxa.html" class="external-link">filter_taxa</a></span><span class="op">(</span><span class="va">ps_sup</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We use the <code>fitModels</code> function to fit all the available
distributions on our dataset:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GOF_plaque_16S</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/fitModels.html" class="external-link">fitModels</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">ps_sup</span>,</span>
<span>    models <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NB"</span>, <span class="st">"ZINB"</span>, <span class="st">"DM"</span>, <span class="st">"ZIG"</span>, <span class="st">"HURDLE"</span><span class="op">)</span>,</span>
<span>    scale_HURDLE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"median"</span><span class="op">)</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># TRUE is always suggested</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="comparing-estimated-and-observed-values">Comparing estimated and observed values<a class="anchor" aria-label="anchor" href="#comparing-estimated-and-observed-values"></a>
</h5>
<p>For each taxon we can compare:</p>
<ul>
<li><p>the Mean Difference (<strong>MD</strong>) <em>i.e.</em> the
difference between the estimated mean and the observed mean abundance
(log scale);</p></li>
<li><p>the Zero Probability Difference (<strong>ZPD</strong>)
<em>i.e.</em> the difference between the probability to observe a zero
and the observed proportion of samples which have zero counts.</p></li>
</ul>
<p>Using the <code>plotMD</code> function we can study the results. What
we are expecting to see are no systematic trends for the mean difference
values.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotMD.html" class="external-link">plotMD</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">GOF_plaque_16S</span>,</span>
<span>    difference <span class="op">=</span> <span class="st">"MD"</span>,</span>
<span>    split <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 23 rows containing non-finite values (stat_smooth).</span></span>
<span><span class="co">#&gt; Warning: Removed 23 rows containing missing values (geom_point).</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/plotGOFMD-1.png" alt="MD plot. Mean-difference (MD) between the estimated and observed count values for each distribution." width="1440"><p class="caption">
MD plot. Mean-difference (MD) between the estimated and observed count
values for each distribution.
</p>
</div>
<p>From the figure above we can see that <em>DM</em> distribution
overestimates the logarithm of the average counts for low values, while
both <em>HURDLE_median</em> and <em>ZIG</em> distributions present an
underestimation as the observed values increase. <em>NB</em> and
<em>ZINB</em> distributions produce very similar estimated and observed
values.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotMD.html" class="external-link">plotMD</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">GOF_plaque_16S</span>,</span>
<span>    difference <span class="op">=</span> <span class="st">"ZPD"</span>,</span>
<span>    split <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/plotGOFZPD-1.png" alt="ZPD plot. Mean-difference between the estimated probability to observe a zero and the observed proportion of zero values (ZPD)." width="1440"><p class="caption">
ZPD plot. Mean-difference between the estimated probability to observe a
zero and the observed proportion of zero values (ZPD).
</p>
</div>
<p>From the figure above we can see that <em>ZIG</em> and <em>NB</em>
models underestimate the probability to observe a zero for sparse
features, while the <em>HURDLE_median</em> model presents a perfect fit
as the probability to observe a zero is the zero rate itself by
construction. <em>DM</em> and <em>ZINB</em> models produce estimated
values very similar to the observed ones.</p>
<p>Overall, in this dataset, <em>NB</em> and <em>ZINB</em> distributions
seems to work well both for the estimation of the average count value
and the zero fraction value. <em>DM</em> model works well for estimating
the probability to observe a zero.</p>
<hr>
<blockquote>
<p><strong>Try it yourself</strong></p>
<p>Try the <code>plotMD</code> function with <em>split = FALSE</em> to
collapse all the smoothing lines on a single plot.</p>
<p>As already mentioned, to summarize the goodness of fit, the Root Mean
Squared Error (RMSE) metric is used. The summary statistics for the
overall performance can be printed using the <code>plotRMSE</code>
function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotRMSE.html" class="external-link">plotRMSE</a></span><span class="op">(</span><span class="va">GOF_plaque_16S</span>, difference <span class="op">=</span> <span class="st">"MD"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotRMSE.html" class="external-link">plotRMSE</a></span><span class="op">(</span><span class="va">GOF_plaque_16S</span>, difference <span class="op">=</span> <span class="st">"ZPD"</span><span class="op">)</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/plotGOFRMSE-1.png" alt="RMSE plot. Root Mean Squared Errors (RMSE) for both the MD and ZPD values." width="960"><p class="caption">
RMSE plot. Root Mean Squared Errors (RMSE) for both the MD and ZPD
values.
</p>
</div>
<p>The lower the RMSE value, the better the goodness of fit of the
model.</p>
</blockquote>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="differential-abundance-detection-methods">Differential Abundance detection methods<a class="anchor" aria-label="anchor" href="#differential-abundance-detection-methods"></a>
</h4>
<p>Before starting, please remember that the data pre-processing,
including QC analysis and filtering steps, are not topics of this
workshop but in real life situations those steps precede the DA analysis
and they are of extreme importance to obtain reliable results.</p>
<p>A list of the available methods is presented below:</p>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<caption>
DA methods available in benchdamic.
</caption>
<thead><tr>
<th style="text-align:left;font-weight: bold;color: black !important;">
Method (package)
</th>
<th style="text-align:left;font-weight: bold;color: black !important;">
Short description
</th>
<th style="text-align:left;font-weight: bold;color: black !important;">
Test
</th>
<th style="text-align:left;font-weight: bold;color: black !important;">
Normalization / Transformation
</th>
<th style="text-align:left;font-weight: bold;color: black !important;">
Suggested input
</th>
<th style="text-align:left;font-weight: bold;color: black !important;">
Output
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_basic (stats)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Simple t and wilcox tests
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
test = ‘t’, ‘wilcox’
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
no normalization
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
all types
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
p-values and statistics
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_edgeR (edgeR)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Negative Binomial (NB) generalized linear model.
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Empirical Bayes + moderated t, robust estimation of priors (if robust =
TRUE), with or without weights (generated by weights_ZINB)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
norm = ‘TMM’, ‘TMMwsp’, ‘RLE’, ‘upperquartile’, ‘posupperquartile’,
‘none’ (produced by norm_edgeR)
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
raw counts with edgeR normalization factors
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
p-values and statistics
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_limma (limma)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Linear model with weights
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Empirical Bayes + moderated t, with or without weights (generated by
weights_ZINB)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
norm = ‘TMM’, ‘TMMwsp’, ‘RLE’, ‘upperquartile’, ‘posupperquartile’,
‘none’ (produced by norm_edgeR)
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
raw counts with edgeR normalization factors
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
p-values and statistics
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_DESeq2 (DESeq2)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Negative Binomial (NB) generalized linear model.
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Empirical Bayes + LRT, with or without weights (generated by
weights_ZINB)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
norm = ‘ratio’, ‘poscounts’, ‘iterate’ (produced by norm_DESeq2)
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
raw counts with DESeq2 size factors
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
p-values and statistics
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_NOISeq (NOISeq)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Non-parametric approach for the comparison of tag-wise statistics to the
noise distribution to detect differential expression
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
M and D statistics compared with the noise distribution
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
norm = ‘rpkm’, ‘uqua’, ‘tmm’, ‘n’
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
raw counts or normalized/transformed counts setting norm = “n”
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
adjusted p-values and statistics
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_dearseq (dearseq)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Variance component score test accounting for data heteroscedasticity
through precision weights
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
test = ‘asymptotic’, ‘permutation’
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
automatically transforms raw counts into log(CPM) (if preprocessed =
FALSE)
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
raw counts or log transformed counts setting preprocessed = TRUE
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
p-values
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_metagenomeSeq (metagenomeSeq)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Zero-Inflated Gaussian (ZIG) mixture model
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
model = ‘fitZig’, ‘fitFeatureModel’
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
norm = ‘CSS’ (produced by norm_CSS)
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
raw counts with CSS normalization factors
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
p-values and statistics
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_corncob (corncob) [temporary unavailable]
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Beta-Binomial regression model
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
test = ‘LRT’, ‘Wald’ with (if boot = TRUE) or without bootstrap
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
automatically transforms raw counts into relative abundances (like using
TSS)
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
raw counts
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
p-values and statistics
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_ALDEx2 (ALDEx2)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Compositional approach - Monte-Carlo simulations from a Dirichlet
distribution
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
test = ‘t’, ‘wilcox’, ‘kw’, ‘ANOVA’, ‘glm’
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
denom = ‘all’, ‘iqlr’, ‘zero’, ‘lvha’, ‘median’, or decided by the user
(with test = ‘glm’, denom = ‘all’)
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
raw counts
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
p-values and statistics
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_ANCOM (ANCOMBC)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Compositional approach - Analysis of microbiome compositions with or
without “sampling fraction” bias correction
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
ANCOM-II algorithm + bias correction (if BC = TRUE) or ANCOM
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
automatic in-method Additive Log Ratio (ALR) like transformation +
normalization based on zero types imputation (if BC = TRUE)
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
raw counts
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
p-values and statistics (if BC = TRUE), only statistics otherwise
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_Seurat (Seurat)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Data transformed and scaled before a test
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
test = ‘wilcox’, ‘bimod’, ‘roc’, ‘t’, ‘negbinom’, ‘poisson’, ‘LR’,
‘MAST’, ‘DESeq2’
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
norm = ‘LogNormalize’, ‘CLR’, ‘RC’, ‘none’ (with test = ‘negbinom’,
‘poisson’, or ‘DESeq2’, data are not transformed). scale.factor
parameter to scale data
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
raw counts
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
p-values and statistics
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; color: black !important;">
DA_MAST (MAST)
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Truncated Gaussian hurdle model
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
Standard Empirical Bayes + LRT
</td>
<td style="text-align:left;width: 6cm; color: black !important;">
log(CPM) transformation + in-method normalization
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
raw counts
</td>
<td style="text-align:left;width: 3cm; color: black !important;">
p-values and statistics
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="type-i-error-control">Type I Error Control<a class="anchor" aria-label="anchor" href="#type-i-error-control"></a>
</h4>
<p><strong>Assumption</strong>: Many methods do not control the number
of false discoveries.</p>
<p><strong>Question</strong>: Which are the differential abundance
methods which can control the number of false positives in your
data?</p>
<hr>
<blockquote>
<p><strong>TIEC structure</strong></p>
<p>Briefly, we randomly assign each sample to one of two experimental
groups (using the <code>createMocks</code> function) and perform DA
analysis (with the <code>runMocks</code> function) between these groups,
repeating the process 10 times (at least 1000 times are suggested). In
this setting, the p-values of a perfect test should be uniformly
distributed between 0 and 1 and the false positive rate (FPR or observed
<span class="math inline">\(\alpha\)</span>), which is the observed
proportion of significant tests, should match the nominal value
(<em>e.g.</em> <span class="math inline">\(\alpha=0.05\)</span>).
Finally, we will use the <code>createTIEC</code> function which produces
the data.frame for plotting the results.</p>
<p>The relationships between the functions used in this section are
explained by the next diagram:</p>
<div class="figure">
<img src="TIEC_structure.svg" alt=""><p class="caption">Type I Error Control diagram.</p>
</div>
</blockquote>
<hr>
<ol style="list-style-type: decimal">
<li><p>Using <code>createMocks</code> function we randomly assign a
label (‘grp1’ or ‘grp2’) to the samples which should be
homogeneous;</p></li>
<li><p>we run DA methods to find differences between the two groups
using <code>runMocks</code>;</p></li>
<li><p>we count the number of DA feature for each method, these are
False Positives by construction;</p></li>
<li><p>we repeat 1-3 many times (<em>N = 10</em>, but at least 1000 is
suggested) and we average the results using the <code>createTIEC</code>
function.</p></li>
</ol>
<div class="section level5">
<h5 id="create-mock-comparisons">Create mock comparisons<a class="anchor" aria-label="anchor" href="#create-mock-comparisons"></a>
</h5>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">my_mocks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/createMocks.html" class="external-link">createMocks</a></span><span class="op">(</span></span>
<span>    nsamples <span class="op">=</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html" class="external-link">nsamples</a></span><span class="op">(</span><span class="va">ps_sup</span><span class="op">)</span>,</span>
<span>    N <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span> <span class="co"># At least N = 1000 is suggested</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="set-up-normalizations-and-da-methods">Set up normalizations and DA methods<a class="anchor" aria-label="anchor" href="#set-up-normalizations-and-da-methods"></a>
</h5>
<p>Firstly, by using the <code>setNormalizations</code> and
<code>runNormalizations</code> methods, we add to the
<code>phyloseq</code> object (or <code>TreeSummarizedExperiment</code>
object) some scaling factors such as <em>TMM</em> from
<code>edgeR</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the instructions</span></span>
<span><span class="va">my_normalizations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/setNormalizations.html" class="external-link">setNormalizations</a></span><span class="op">(</span></span>
<span>    fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"norm_edgeR"</span><span class="op">)</span>, </span>
<span>    method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TMM"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add the normalization/scaling factors to the object</span></span>
<span><span class="va">ps_sup</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/runNormalizations.html" class="external-link">runNormalizations</a></span><span class="op">(</span></span>
<span>  normalization_list <span class="op">=</span> <span class="va">my_normalizations</span>, </span>
<span>  object <span class="op">=</span> <span class="va">ps_sup</span>, </span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;       + Running now:norm_edgeR</span></span>
<span><span class="co">#&gt;         Parameters:method=TMM</span></span>
<span><span class="co">#&gt; Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span></span>
<span><span class="co">#&gt; Also defined by 'tidytree'</span></span>
<span><span class="co">#&gt; Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span></span>
<span><span class="co">#&gt; Also defined by 'tidytree'</span></span>
<span><span class="co">#&gt; Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span></span>
<span><span class="co">#&gt; Also defined by 'tidytree'</span></span>
<span><span class="co">#&gt; Found more than one class "phylo" in cache; using the first, from namespace 'phyloseq'</span></span>
<span><span class="co">#&gt; Also defined by 'tidytree'</span></span>
<span><span class="co">#&gt; NF.TMM column has been added.</span></span></code></pre></div>
<p>Some messages “Found more than one”phylo” class in cache…” could be
shown after running the previous functions. They are caused by
duplicated class names between <code>phyloseq</code> and
<code>tidytree</code> packages and can be ignored.</p>
<hr>
<blockquote>
<p><strong>Try it yourself</strong></p>
<p>Normalization step can also be done manually, for example by using
the <code>norm_edgeR</code>, <code>norm_DESeq2</code>, and
<code>norm_CSS</code> methods.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps_sup</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/norm_edgeR.html" class="external-link">norm_edgeR</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">ps_sup</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"TMM"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ps_sup</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/norm_DESeq2.html" class="external-link">norm_DESeq2</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">ps_sup</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"poscounts"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ps_sup</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/norm_CSS.html" class="external-link">norm_CSS</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">ps_sup</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"CSS"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># See the normalization/scaling factors inside the object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">ps_sup</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NF.TMM"</span>, <span class="st">"NF.poscounts"</span>, <span class="st">"NF.CSS"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</blockquote>
<hr>
<p>After the normalization/scaling factors have been added to the
<code>phyloseq</code> or <code>TreeSummarizedExperiment</code> object,
you could decide to filter rare taxa which do not carry much
information. In this example a simple filter is applied to keep only
features with a count in at least 3 samples.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps_sup_filtered</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/filter_taxa.html" class="external-link">filter_taxa</a></span><span class="op">(</span></span>
<span>    physeq <span class="op">=</span> <span class="va">ps_sup</span>, </span>
<span>    flist <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">3</span>, </span>
<span>    prune <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ps_sup_filtered</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 44 taxa and 30 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 30 samples by 8 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 44 taxa by 6 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; phy_tree()    Phylogenetic Tree: [ 44 tips and 43 internal nodes ]</span></span></code></pre></div>
<p>We set the instructions of a bunch of DA methods. In this workshop we
will use:</p>
<ul>
<li><p><code>limma</code> with <em>TMM</em> scaling factors;</p></li>
<li><p><code>ALDEx2</code> with <em>denom = “all”</em> data
transformation performing the wilcox test;</p></li>
</ul>
<p>Other methods and many combination of parameters are still possible
but not included in this workshop.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/set_limma.html" class="external-link">set_limma</a></span><span class="op">(</span></span>
<span>    pseudo_count <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    design <span class="op">=</span> <span class="op">~</span> <span class="va">group</span>,</span>
<span>    coef <span class="op">=</span> <span class="st">"groupgrp2"</span>,</span>
<span>    norm <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<span>    weights_logical <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    expand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_ALDEx2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/set_ALDEx2.html" class="external-link">set_ALDEx2</a></span><span class="op">(</span></span>
<span>    pseudo_count <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    design <span class="op">=</span> <span class="st">"group"</span>,</span>
<span>    mc.samples <span class="op">=</span> <span class="fl">128</span>,</span>
<span>    test <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>    paired.test <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    denom <span class="op">=</span> <span class="st">"all"</span>, </span>
<span>    contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"group"</span>, <span class="st">"grp2"</span>, <span class="st">"grp1"</span><span class="op">)</span>,</span>
<span>    expand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_methods</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">my_limma</span>, <span class="va">my_ALDEx2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="add-a-custom-method">Add a custom method<a class="anchor" aria-label="anchor" href="#add-a-custom-method"></a>
</h5>
<p>Let’s create a very simple custom DA method.</p>
<p>I want to start from the log-transformed counts (adding 1 to all
counts to avoid log(0)) and perform a t-test.</p>
<p>If you want to try something else, feel free to do it, just remember
to:</p>
<ul>
<li><p>include a <em>verbose = TRUE</em> (or <em>FALSE</em>) parameter
to let the user know what the method is doing;</p></li>
<li><p>return a <em>pValMat</em> matrix which contains the raw p-values
and adjusted p-values in <em>rawP</em> and <em>adjP</em> columns
respectively;</p></li>
<li><p>return a <em>statInfo</em> matrix which contains the summary
statistics for each feature, such as the <em>logFC</em>, standard
errors, test statistics and so on;</p></li>
<li><p>return a <em>name</em> which contains the complete name of the
used method.</p></li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DA_custom</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span></span>
<span>    <span class="va">object</span>, </span>
<span>    <span class="va">contrast</span> <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>    <span class="va">paired</span> <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    <span class="va">verbose</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="op">{</span></span>
<span>    <span class="va">name</span> <span class="op">&lt;-</span> <span class="st">"custom"</span>  </span>
<span>  </span>
<span>    <span class="co"># Read the counts</span></span>
<span>    <span class="va">counts_and_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/get_counts_metadata.html" class="external-link">get_counts_metadata</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span>    <span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">counts_and_metadata</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">counts_and_metadata</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span>    </span>
<span>    <span class="co"># Log transform data</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="va">verbose</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Log transform the counts"</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Check if contrast vector is present</span></span>
<span>    <span class="co"># We need it to understand what to test</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">contrast</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"Please supply a contrast vector"</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="co"># Look for samples of the first and second group</span></span>
<span>    <span class="va">group1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">[</span>, <span class="va">contrast</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="va">contrast</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="va">group2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">[</span>, <span class="va">contrast</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="va">contrast</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="va">paired</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="co"># Check if the number of samples in the groups is equal</span></span>
<span>        <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">group1</span><span class="op">)</span> <span class="op">!=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">group2</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>            <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="va">method</span>, <span class="st">"\n"</span>, <span class="st">"paired = TRUE but "</span>, </span>
<span>                <span class="va">contrast</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">" and "</span>, <span class="va">contrast</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="st">" have "</span>, </span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">group1</span><span class="op">)</span>, <span class="st">" and "</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">group2</span><span class="op">)</span>, </span>
<span>                <span class="st">" samples respectively."</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>        <span class="kw">if</span><span class="op">(</span><span class="va">verbose</span><span class="op">)</span><span class="op">{</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Performing the analysis on paired data"</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>        <span class="co"># Update the name</span></span>
<span>        <span class="va">name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">name</span>, <span class="st">"."</span>, <span class="st">"paired"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="va">verbose</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Performing t.tests"</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="co"># Perform the tests</span></span>
<span>    <span class="va">statInfo_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span></span>
<span>        X <span class="op">=</span> <span class="va">logcounts</span>, </span>
<span>        MARGIN <span class="op">=</span> <span class="fl">1</span>, </span>
<span>        FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">feature</span><span class="op">)</span><span class="op">{</span></span>
<span>            <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">paired</span><span class="op">)</span><span class="op">{</span></span>
<span>                <span class="co"># Compute the average values for each group</span></span>
<span>                <span class="va">avg_group1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">feature</span><span class="op">[</span><span class="va">group1</span><span class="op">]</span><span class="op">)</span> </span>
<span>                <span class="va">avg_group2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">feature</span><span class="op">[</span><span class="va">group2</span><span class="op">]</span><span class="op">)</span></span>
<span>                <span class="co"># Compute the logFC</span></span>
<span>                <span class="va">logFC</span> <span class="op">&lt;-</span> <span class="va">avg_group1</span> <span class="op">-</span> <span class="va">avg_group2</span></span>
<span>            <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>                <span class="co"># Compute the average difference for each group</span></span>
<span>                <span class="va">logFC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">feature</span><span class="op">[</span><span class="va">group1</span><span class="op">]</span> <span class="op">-</span> <span class="va">feature</span><span class="op">[</span><span class="va">group2</span><span class="op">]</span><span class="op">)</span></span>
<span>            <span class="op">}</span></span>
<span>            </span>
<span>            <span class="co"># Perform the t-test</span></span>
<span>            <span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html" class="external-link">t.test</a></span><span class="op">(</span></span>
<span>                x <span class="op">=</span> <span class="va">feature</span><span class="op">[</span><span class="va">group1</span><span class="op">]</span>, </span>
<span>                y <span class="op">=</span> <span class="va">feature</span><span class="op">[</span><span class="va">group2</span><span class="op">]</span>,</span>
<span>                paired <span class="op">=</span> <span class="va">paired</span><span class="op">)</span></span>
<span>            <span class="co"># Prepare the output</span></span>
<span>            <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>                <span class="va">logFC</span>,</span>
<span>                <span class="st">"statistic"</span> <span class="op">=</span> <span class="va">results</span><span class="op">[[</span><span class="st">"statistic"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                <span class="st">"pvalue"</span> <span class="op">=</span> <span class="va">results</span><span class="op">[[</span><span class="st">"p.value"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                <span class="st">"lowCI"</span> <span class="op">=</span> <span class="va">results</span><span class="op">[[</span><span class="st">"conf.int"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                <span class="st">"uppCI"</span> <span class="op">=</span> <span class="va">results</span><span class="op">[[</span><span class="st">"conf.int"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span>            <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span> </span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Transform the list into a data.frame</span></span>
<span>    <span class="va">statInfo</span> <span class="op">&lt;-</span> <span class="fu">plyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyr/man/ldply.html" class="external-link">ldply</a></span><span class="op">(</span><span class="va">statInfo_list</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Rename the columns</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">statInfo</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"taxon"</span></span>
<span>    </span>
<span>    <span class="co"># Create pValMat</span></span>
<span>    <span class="va">padj</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span>p <span class="op">=</span> <span class="va">statInfo</span><span class="op">$</span><span class="va">pvalue</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span>    <span class="va">pValMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"rawP"</span> <span class="op">=</span> <span class="va">statInfo</span><span class="op">[</span>, <span class="st">"pvalue"</span><span class="op">]</span>, <span class="st">"adjP"</span> <span class="op">=</span> <span class="va">padj</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Add rownames</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">statInfo</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pValMat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">statInfo</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span>    </span>
<span>    <span class="co"># Return statInfo, pValMat and a name</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"pValMat"</span> <span class="op">=</span> <span class="va">pValMat</span>, </span>
<span>        <span class="st">"statInfo"</span> <span class="op">=</span> <span class="va">statInfo</span>,</span>
<span>        <span class="st">"name"</span> <span class="op">=</span> <span class="va">name</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="co"># END - function: DA_custom</span></span></code></pre></div>
<p>We add it to the framework by creating a list containing one or more
instances of the custom method with the desired combination of
parameters:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_custom_method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    customMethod.1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        method <span class="op">=</span> <span class="st">"DA_custom"</span>, </span>
<span>        contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"group"</span>, <span class="st">"grp2"</span>, <span class="st">"grp1"</span><span class="op">)</span>,</span>
<span>        paired <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <em>method</em> field, containing the name of the method to call,
is mandatory, while the <em>verbose</em> parameter is not needed.</p>
<p>We put all the instructions together by concatenating them in a
single object called <em>my_methods</em> and then we run them using the
<code>runMocks</code> function:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add the custom method instances to the others</span></span>
<span><span class="va">my_methods</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">my_methods</span>, <span class="va">my_custom_method</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Random grouping each time</span></span>
<span><span class="va">sup_mockDA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/runMocks.html" class="external-link">runMocks</a></span><span class="op">(</span></span>
<span>    mocks <span class="op">=</span> <span class="va">my_mocks</span>, </span>
<span>    method_list <span class="op">=</span> <span class="va">my_methods</span>,</span>
<span>    object <span class="op">=</span> <span class="va">ps_sup_filtered</span>, </span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>If some warnings are reported set <em>verbose = TRUE</em> to obtain
the method name and the mock comparison where the warnings occured.</p>
<p>The structure of the output in this example is the following:</p>
<ul>
<li>
<p><em>Comparison1</em> to <em>Comparison10</em> on the first level,
which contains:</p>
<ul>
<li>
<p><em>Method1</em> to <em>Method3</em> output lists on the second
level:</p>
<ul>
<li><p><em>pValMat</em> which contains the matrix of raw p-values and
adjusted p-values in <em>rawP</em> and <em>adjP</em> columns
respectively;</p></li>
<li><p><em>statInfo</em> which contains the matrix of summary statistics
for each feature, such as the <em>logFC</em>, standard errors, test
statistics and so on;</p></li>
<li><p><em>name</em> which contains the complete name of the used
method.</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="counting-the-false-positives">Counting the False Positives<a class="anchor" aria-label="anchor" href="#counting-the-false-positives"></a>
</h5>
<p>We run the <code>createTIEC</code> function for counting the False
Positives:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TIEC_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/createTIEC.html" class="external-link">createTIEC</a></span><span class="op">(</span><span class="va">sup_mockDA</span><span class="op">)</span></span></code></pre></div>
<p>We can plot the results by using the <code>plotFPR</code>
function:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/createColors.html" class="external-link">createColors</a></span><span class="op">(</span></span>
<span>  variable <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">TIEC_summary</span><span class="op">$</span><span class="va">df_FPR</span><span class="op">$</span><span class="va">Method</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotFPR.html" class="external-link">plotFPR</a></span><span class="op">(</span>df_FPR <span class="op">=</span> <span class="va">TIEC_summary</span><span class="op">$</span><span class="va">df_FPR</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using Comparison, Method as id variables</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/FPRplot-1.png" alt="FPR plot. Boxplots of the proportion of raw p-values lower than the commonly used thresholds for the nominal $\alpha$ (i.e. the False Positive Rate) for each DA method." width="768"><p class="caption">
FPR plot. Boxplots of the proportion of raw p-values lower than the
commonly used thresholds for the nominal <span class="math inline">\(\alpha\)</span> (i.e. the False Positive Rate) for
each DA method.
</p>
</div>
<p>The false positive rate (FPR or observed <span class="math inline">\(\alpha\)</span>), which is the observed proportion
of significant tests, should match the nominal value (red dotted lines)
because all the findings are false positive by construction. In this
example <code>ALDEx2.all.wilcox.unpaired</code> has a conservative
behavior as it is below the thresholds. Instead, <code>limma.TMM</code>
seems to be conservative for the lowest threshold and in line with the
expectations for 0.05 and 0.1 thresholds.</p>
<hr>
<blockquote>
<p><strong>Other graphical representations</strong></p>
<p><code>createTIEC</code> produces a list of 4 data frames and a
graphical representation is available for each of them:</p>
<ol style="list-style-type: decimal">
<li><p><em>df_pval</em> is a 5 columns and <em>number_of_features x
methods x comparisons</em> rows data frame. The five columns are called
<em>Comparison</em>, <em>Method</em>, <em>variable</em> (which contains
the feature names), <em>pval</em> and <em>padj</em>;</p></li>
<li><p><em>df_FPR</em> is a 5 columns and <em>methods x comparisons</em>
rows data frame. For each set of method and comparison, the proportion
of false positives, considering 3 threshold (0.01, 0.05, 0.1) are
reported;</p></li>
<li><p><em>df_QQ</em> contains the average p-value for each theoretical
quantile, <em>i.e.</em> the coordinates to draw the QQ-plot for
comparing the mean observed p-values distribution across comparisons,
with the theoretical uniform distribution. Indeed, the observed p-values
should follow a uniform distribution under the null hypothesis of no
differential abundant features presence;</p></li>
<li><p><em>df_KS</em> is a 5 columns and <em>methods x comparisons</em>
rows data frame. For each set of method and comparison, the
Kolmogorov-Smirnov test statistics and pvalues are reported in
<em>KS</em> and <em>KS_pval</em> columns respectively.</p></li>
</ol>
<p><strong>QQ-Plot</strong></p>
<p>The p-values distribution under the null hypothesis should be
uniform. This is qualitatively summarized in the QQ-plot.</p>
<p>Methods over the bisector show a conservative behavior, while methods
below the bisector a liberal one:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotQQ.html" class="external-link">plotQQ</a></span><span class="op">(</span>df_QQ <span class="op">=</span> <span class="va">TIEC_summary</span><span class="op">$</span><span class="va">df_QQ</span>, zoom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, cols <span class="op">=</span> <span class="va">cols</span>, split <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/QQplotsplit-1.png" alt="QQ plot. Quantile-quantile plots from 0 to 1 for each DA method are displayed separately. Average curves are reported" width="960"><p class="caption">
QQ plot. Quantile-quantile plots from 0 to 1 for each DA method are
displayed separately. Average curves are reported
</p>
</div>
<p><strong>Kolmogorov-Smirnov test</strong></p>
<p>Departure from uniformity is quantitatively evaluated through the
Kolmogorov-Smirnov test:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotKS.html" class="external-link">plotKS</a></span><span class="op">(</span>df_KS <span class="op">=</span> <span class="va">TIEC_summary</span><span class="op">$</span><span class="va">df_KS</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/KSplot-1.png" alt="KS plot. Kolmogorov-Smirnov (KS) statistic boxplots for each DA methods where the raw p-values distribution is compared with a uniform distribution." width="672"><p class="caption">
KS plot. Kolmogorov-Smirnov (KS) statistic boxplots for each DA methods
where the raw p-values distribution is compared with a uniform
distribution.
</p>
</div>
<p>High KS values indicates departure from the uniformity while low
values indicates closeness. <code>ADLEx2.all.wilcox.unpaired</code>,
which was very conservative has distribution of p-values far from
uniformity. <code>custom</code> method instead has the lowest KS
value.</p>
<p><strong>Log distribution of p-values</strong></p>
<p>Looking at the p-values’ log-scale can also be informative. This is
because behavior in the tail may be poor even when the overall p-values
distribution is uniform, with a few unusually small p-values in an
otherwise uniform distribution.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotLogP.html" class="external-link">plotLogP</a></span><span class="op">(</span>df_QQ <span class="op">=</span> <span class="va">TIEC_summary</span><span class="op">$</span><span class="va">df_QQ</span>, cols <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/AveLogPplot-1.png" alt="-log10(average p-value) plot ('average' refers to the average p-value computed for each quantile across mocks comparisons). Negative logarithm distribution of average p-values. Red-shaded vertical bars represent the 90, 95, and 99 percentiles for the negative log distribution of average p-values for each method. They should align with the dotted lines which represent the percentiles of the IDEAL distribution." width="672"><p class="caption">
-log10(average p-value) plot (‘average’ refers to the average p-value
computed for each quantile across mocks comparisons). Negative logarithm
distribution of average p-values. Red-shaded vertical bars represent the
90, 95, and 99 percentiles for the negative log distribution of average
p-values for each method. They should align with the dotted lines which
represent the percentiles of the IDEAL distribution.
</p>
</div>
<p>The <span class="math inline">\(-\log_{10}(p-value)\)</span> IDEAL
distribution is reported in red color as the first method. To highlight
tail’s behaviors, 3 percentiles (0.9, 0.95, 0.99) are reported using
red-shaded vertical segments for each method. If the method’s
distribution of negative log-transformed p-values or average p-values is
still uniform in the 3 selected quantiles of the tail, the 3 red
vertical segments will align to the respective dotted line. Usually,
when a method has its red segments to the left of the IDEAL’s
(<em>e.g.</em> <code>ALDEx2.all.wilcox.unpaired</code>) its behavior is
conservative. Indeed, for those methods, little p-values are fewer than
expected. On the contrary, methods with red segments to the right of the
IDEAL’s have a liberal behavior. Overall, <code>limma.TMM</code> has the
best performances among the three methods.</p>
</blockquote>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="concordance">Concordance<a class="anchor" aria-label="anchor" href="#concordance"></a>
</h4>
<p><strong>Assumption</strong>: Applying different methods to the same
data may produce different results.</p>
<p><strong>Questions</strong>: How much do the methods agree with each
other? How much does a method agree with itself?</p>
<hr>
<blockquote>
<p><strong>Concordance structure</strong></p>
<p>Several functions are involved in the concordance analysis. First of
all, the normalization methods need to be set up using the
<code>setNormalizations</code> function. The differential abundance
framework is set up by using <code>set_*</code> functions. Once all
these instructions are set up, we can call the <code>runSplits</code>
function. Finally, the <code>createConcordance</code> and
<code>plotConcordance</code> functions are used to compute and plot the
results.</p>
<p>The relationships between the functions used in this section are
explained by the next diagram:</p>
<div class="figure">
<img src="concordance_structure.svg" alt=""><p class="caption">Concordance analysis diagram.</p>
</div>
</blockquote>
<hr>
<p>If we want to measure the ability of each method to produce
replicable results:</p>
<ol style="list-style-type: decimal">
<li><p>we consider a dataset with two or more groups and split it by
half to obtain the Subset1 and Subset2 datasets using the
<code>createSplits</code> function;</p></li>
<li><p>we run DA methods on both subsets using the
<code>runSplits</code> function;</p></li>
<li><p>we compute the Concordance At the Top metric (CAT) between the
lists of p-values to obtain the Between Method Concordances (BMC) and
the Within Method Concordances (WMC);</p></li>
<li><p>we repeat the 1-3 many times (<em>N = 10</em>, but at least 100
is suggested) and we average the results using the
<code>createConcordance</code> function.</p></li>
</ol>
<p>This time we consider the samples from both the supragingival and
subgingival plaques.</p>
<div class="section level5">
<h5 id="split-datasets">Split datasets<a class="anchor" aria-label="anchor" href="#split-datasets"></a>
</h5>
<p>We split the dataset considering the paired design:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the HMP_BODY_SUBSITE and RSID variable as factors</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">ps_plaque_16S</span><span class="op">)</span><span class="op">$</span><span class="va">HMP_BODY_SUBSITE</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">ps_plaque_16S</span><span class="op">)</span><span class="op">$</span><span class="va">HMP_BODY_SUBSITE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">ps_plaque_16S</span><span class="op">)</span><span class="op">$</span><span class="va">RSID</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">ps_plaque_16S</span><span class="op">)</span><span class="op">$</span><span class="va">RSID</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set a seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">my_splits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/createSplits.html" class="external-link">createSplits</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">ps_plaque_16S</span>,</span>
<span>  varName <span class="op">=</span> <span class="st">"HMP_BODY_SUBSITE"</span>,</span>
<span>  paired <span class="op">=</span> <span class="st">"RSID"</span>,</span>
<span>  balanced <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  N <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span> <span class="co"># At least 100 is suggested</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="set-up-normalizations-and-da-methods-1">Set up normalizations and DA methods<a class="anchor" aria-label="anchor" href="#set-up-normalizations-and-da-methods-1"></a>
</h5>
<p>For some of the methods implemented in this package it is possible to
perform differential abundance testings for the repeated measurements
experimental designs (<em>e.g.</em> by adding the subject ID in the
model formula of <code>limma</code>).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/set_limma.html" class="external-link">set_limma</a></span><span class="op">(</span></span>
<span>    design <span class="op">=</span> <span class="op">~</span> <span class="fl">1</span> <span class="op">+</span> <span class="va">RSID</span> <span class="op">+</span> <span class="va">HMP_BODY_SUBSITE</span>,</span>
<span>    coef <span class="op">=</span> <span class="st">"HMP_BODY_SUBSITESupragingival Plaque"</span>, </span>
<span>    norm <span class="op">=</span> <span class="st">"TMM"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_ALDEx2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/set_ALDEx2.html" class="external-link">set_ALDEx2</a></span><span class="op">(</span></span>
<span>    pseudo_count <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    design <span class="op">=</span> <span class="st">"HMP_BODY_SUBSITE"</span>,</span>
<span>    mc.samples <span class="op">=</span> <span class="fl">128</span>,</span>
<span>    test <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>    paired.test <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    denom <span class="op">=</span> <span class="st">"all"</span>, </span>
<span>    contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"HMP_BODY_SUBSITE"</span>, </span>
<span>        <span class="st">"Supragingival Plaque"</span>, </span>
<span>        <span class="st">"Subgingival Plaque"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_custom_method</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    customMethod.1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        method <span class="op">=</span> <span class="st">"DA_custom"</span>, </span>
<span>        contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>            <span class="st">"HMP_BODY_SUBSITE"</span>, </span>
<span>            <span class="st">"Supragingival Plaque"</span>, </span>
<span>            <span class="st">"Subgingival Plaque"</span><span class="op">)</span>,</span>
<span>        paired <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_methods</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="va">my_limma</span>,  </span>
<span>    <span class="va">my_ALDEx2</span>, </span>
<span>    <span class="va">my_custom_method</span><span class="op">)</span></span></code></pre></div>
<p>Similarly, to set the normalization methods, the
<code>setNormalizations</code> function can be used. In this case it has
already been set up for the TIEC analysis:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">my_normalizations</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 1</span></span>
<span><span class="co">#&gt;  $ norm_edgeR:List of 2</span></span>
<span><span class="co">#&gt;   ..$ fun   : chr "norm_edgeR"</span></span>
<span><span class="co">#&gt;   ..$ method: chr "TMM"</span></span></code></pre></div>
<p>The <code>runSplits</code> function generates the subsets and
performs DA analysis on the features with at least 1 (min_counts &gt; 0)
count in more than 2 samples (min_samples &gt; 2):</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Plaque_16S_splitsDA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/runSplits.html" class="external-link">runSplits</a></span><span class="op">(</span></span>
<span>    split_list <span class="op">=</span> <span class="va">my_splits</span>, </span>
<span>    method_list <span class="op">=</span> <span class="va">my_methods</span>, </span>
<span>    normalization_list <span class="op">=</span> <span class="va">my_normalizations</span>, </span>
<span>    object <span class="op">=</span> <span class="va">ps_plaque_16S</span>, </span>
<span>    min_counts <span class="op">=</span> <span class="fl">0</span>, min_samples <span class="op">=</span> <span class="fl">2</span>, </span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The structure of the output in this example is the following:</p>
<ul>
<li>
<p><em>Subset1</em> and <em>Subset2</em> on the first level, which
contains:</p>
<ul>
<li>
<p><em>Comparison1</em> to <em>Comparison10</em> output lists on the
second level:</p>
<ul>
<li>
<p>results of the 3 methods on the third level: <code>limma</code>
with <em>TMM</em> scaling factors, <code>ALDEx2</code> with paired
<em>wilcox</em> test and denom equals to <em>all</em>, and
<code>custom</code> for paired data. They are organized as already
described:</p>
<ul>
<li><p><em>pValMat</em> which contains the matrix of raw p-values and
adjusted p-values in <em>rawP</em> and <em>adjP</em> columns
respectively;</p></li>
<li><p><em>statInfo</em> which contains the matrix of summary statistics
for each feature, such as the <em>logFC</em>, standard errors, test
statistics and so on;</p></li>
<li><p><em>name</em> which contains the complete name of the used
method.</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level5">
<h5 id="comparing-the-concordances">Comparing the concordances<a class="anchor" aria-label="anchor" href="#comparing-the-concordances"></a>
</h5>
<p>For each pair of methods the concordance is computed by the
<code>createConcordance</code> function:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">concordance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/createConcordance.html" class="external-link">createConcordance</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">Plaque_16S_splitsDA</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
<blockquote>
<p><strong>More details</strong></p>
<p>The <code>createConcordance</code> function produces a long format
data frame object with several columns:</p>
<ul>
<li>
<em>comparison</em> which indicates the comparison number;</li>
<li>
<em>n_features</em> which indicates the total number of taxa in the
comparison dataset;</li>
<li>name of <em>method1</em>;</li>
<li>name of <em>method2</em>;</li>
<li>
<em>rank</em>;</li>
<li>
<em>concordance</em> which is defined as the cardinality of the
intersection of the top <em>rank</em> elements of each list, divided by
<em>rank</em>, <em>i.e.</em>, <span class="math inline">\(\frac{L_{1:rank} \bigcap
M_{1:rank}}{rank}\)</span>, where <em>L</em> and <em>M</em> represent
the lists of p-values of <em>method1</em> and <em>method2</em>
respectively. A noise value (<span class="math inline">\(&lt;10^{-10}\)</span>) is added to each p-value
(or statistic) in order to avoid duplicated values which cannot be
ordered.</li>
</ul>
<p><strong>Concordances based on other statistics</strong></p>
<p>The <code>createConcordance</code> method is very flexible. In the
example below the concordances are built using the log fold changes or
other statistics instead of the p-values. To do so, it is necessary to
know the column names generated by each differential abundance method in
the <em>statInfo</em> matrix.</p>
<p>Firstly, inspect the method order:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Plaque_16S_splitsDA</span><span class="op">$</span><span class="va">Subset1</span><span class="op">$</span><span class="va">Comparison1</span><span class="op">)</span></span></code></pre></div>
<p>Then, look at the column names:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"limma.TMM"</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Plaque_16S_splitsDA</span><span class="op">$</span><span class="va">Subset1</span><span class="op">$</span><span class="va">Comparison1</span><span class="op">$</span><span class="va">limma.TMM</span><span class="op">$</span><span class="va">statInfo</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"ALDEx2.all.wilcox.paired"</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Plaque_16S_splitsDA</span><span class="op">$</span><span class="va">Subset1</span><span class="op">$</span><span class="va">Comparison1</span><span class="op">$</span><span class="va">ALDEx2.all.wilcox.paired</span><span class="op">$</span></span>
<span>    <span class="va">statInfo</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"custom.paired"</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Plaque_16S_splitsDA</span><span class="op">$</span><span class="va">Subset1</span><span class="op">$</span><span class="va">Comparison1</span><span class="op">$</span><span class="va">custom.paired</span><span class="op">$</span><span class="va">statInfo</span><span class="op">)</span></span></code></pre></div>
<p>All methods, except for ALDEx2, contain the log fold change values in
the <em>logFC</em> column of <em>statInfo</em> matrix. Knowing this, the
concordance data frame can be built using:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">concordance_alternative</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/createConcordance.html" class="external-link">createConcordance</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">Plaque_16S_splitsDA</span>,</span>
<span>    slot <span class="op">=</span> <span class="st">"statInfo"</span>,</span>
<span>    colName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"effect"</span>, <span class="st">"logFC"</span><span class="op">)</span>,</span>
<span>    type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"logfc"</span>, <span class="st">"logfc"</span>, <span class="st">"logfc"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</blockquote>
<hr>
<p>Using the <code>plotConcordance</code> function the CAT values from
rank 1 to 30 are drawn. The area between the curve and the bisector is
colored to highlight concordant methods (blue) and non-concordant ones
(red).</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotConcordance.html" class="external-link">plotConcordance</a></span><span class="op">(</span></span>
<span>    concordance <span class="op">=</span> <span class="va">concordance</span>, </span>
<span>    threshold <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>    plotlist <span class="op">=</span> <span class="va">pC</span>, </span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span>, </span>
<span>    align <span class="op">=</span> <span class="st">"h"</span>, </span>
<span>    axis <span class="op">=</span> <span class="st">"tb"</span>,</span>
<span>    rel_widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/plotConcordance-1.png" alt="BMC and WMC plot. Between-method concordance (BMC) and within-method concordance (WMC) (main diagonal) averaged values from rank 1 to 30." width="768"><p class="caption">
BMC and WMC plot. Between-method concordance (BMC) and within-method
concordance (WMC) (main diagonal) averaged values from rank 1 to 30.
</p>
</div>
<p><code>limma.TMM</code> and <code>custom</code> methods show the
highest BMC values while <code>ALDEx2.all.wilcox.paired</code> shows the
highest WMC value.</p>
<hr>
<blockquote>
<p><strong>More about WMC and BMC metrics</strong></p>
<p>It is common that WMC values (in red rectangles) are lower than BMC
ones. Indeed, BMC is computed between different methods on the same
Subset, while WMC is computed for a single method between the results
obtained in Subset1 and Subset2 (some taxa are dataset-specific).</p>
<p>Mock comparisons and random splits allow to evaluate type I error and
concordance. However, these analyses do not assess the correctness of
the discoveries. Even the method with the highest WMC could nonetheless
consistently identify false positive DA taxa.</p>
</blockquote>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="enrichment-analysis">Enrichment analysis<a class="anchor" aria-label="anchor" href="#enrichment-analysis"></a>
</h4>
<p><strong>Assumption</strong>: Previous analyses did not assess the
correctness of the discoveries.</p>
<p><strong>Question</strong>: If some prior knowledge about your
experiment is available, would the finding be coherent with that
knowledge?</p>
<hr>
<blockquote>
<p><strong>Enrichment analysis structure</strong></p>
<p>After performing DA analysis on a dataset where some prior knowledge
is given, <code>createEnrichment</code> and <code>createPositives</code>
functions can be used to analyse the enrichment of DA features in
specific levels of the prior knowledge variable.</p>
<p>The relationships between the functions used in this section are
explained by the next diagram:</p>
<div class="figure">
<img src="enrichment_structure.svg" alt=""><p class="caption">Enrichment analysis diagram.</p>
</div>
</blockquote>
<hr>
<ul>
<li><p>We run DA methods to test a difference between two experimental
groups using the <code>runDA</code> function;</p></li>
<li><p>we use the prior knowledge (if present) to check the correctness
of the findigs using the <code>createEnrichment</code>
function.</p></li>
</ul>
<div class="section level5">
<h5 id="a-priori-knowledge">
<em>A priori</em> knowledge<a class="anchor" aria-label="anchor" href="#a-priori-knowledge"></a>
</h5>
<p>Here, we leveraged the peculiar environment of the gingival site:</p>
<ul>
<li><p>the <strong>supragingival biofilm</strong> is directly exposed to
the open atmosphere of the oral cavity, favoring the growth of aerobic
species;</p></li>
<li><p>In the <strong>subgingival biofilm</strong>, the atmospheric
conditions gradually become strict anaerobic, favoring the growth of
anaerobic species.</p></li>
</ul>
<p>From the comparison of the two sites, we thus expect to find an
abundance of <strong>aerobic microbes in the supragingival
plaque</strong> and of <strong>anaerobic bacteria in the subgingival
plaque</strong>.</p>
<p>Microbial metabolism data comes from a research article titled
“Tobacco exposure associated with oral microbiota oxygen utilization in
the New York City Health and Nutrition Examination Study” by <a href="https://doi.org/10.1016/j.annepidem.2019.03.005" class="external-link">Beghini et al.
(2019)</a>. They can be loaded using
<code>data("microbial_metabolism")</code>:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"microbial_metabolism"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">microbial_metabolism</span><span class="op">)</span></span>
<span><span class="co">#&gt;              Genus        Type</span></span>
<span><span class="co">#&gt; 1     Acholeplasma F Anaerobic</span></span>
<span><span class="co">#&gt; 2 Actinomycetaceae F Anaerobic</span></span>
<span><span class="co">#&gt; 3    Aeriscardovia     Aerobic</span></span>
<span><span class="co">#&gt; 4       Aerococcus F Anaerobic</span></span>
<span><span class="co">#&gt; 5  Aggregatibacter F Anaerobic</span></span>
<span><span class="co">#&gt; 6    Alloscardovia   Anaerobic</span></span></code></pre></div>
<p>We have to match each taxon of the phyloseq object to its type of
metabolism:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract genera from the phyloseq tax_table slot</span></span>
<span><span class="va">genera</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">ps_plaque_16S</span><span class="op">)</span><span class="op">[</span>, <span class="st">"GENUS"</span><span class="op">]</span></span>
<span><span class="co"># Genera as rownames of microbial_metabolism data.frame</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">microbial_metabolism</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">microbial_metabolism</span><span class="op">$</span><span class="va">Genus</span></span>
<span></span>
<span><span class="co"># Match OTUs to their metabolism</span></span>
<span><span class="va">priorInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">genera</span>, </span>
<span>    <span class="st">"Type"</span> <span class="op">=</span>  <span class="va">microbial_metabolism</span><span class="op">[</span><span class="va">genera</span>, <span class="st">"Type"</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Populate the Unknowns</span></span>
<span><span class="va">unknown_metabolism</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">priorInfo</span><span class="op">$</span><span class="va">Type</span><span class="op">)</span></span>
<span><span class="va">priorInfo</span><span class="op">[</span><span class="va">unknown_metabolism</span>, <span class="st">"Type"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Unknown"</span></span>
<span></span>
<span><span class="co"># Relabel 'F Anaerobic' to 'F_Anaerobic' to remove space</span></span>
<span><span class="va">priorInfo</span><span class="op">$</span><span class="va">Type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span></span>
<span>    <span class="va">priorInfo</span><span class="op">$</span><span class="va">Type</span>, </span>
<span>    levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Aerobic"</span>,<span class="st">"Anaerobic"</span>,<span class="st">"F Anaerobic"</span>,<span class="st">"Unknown"</span><span class="op">)</span>, </span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Aerobic"</span>,<span class="st">"Anaerobic"</span>,<span class="st">"F_Anaerobic"</span>,<span class="st">"Unknown"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add a more informative names column</span></span>
<span><span class="va">priorInfo</span><span class="op">[</span>, <span class="st">"newNames"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">priorInfo</span><span class="op">)</span>, <span class="st">"|"</span>,</span>
<span>    <span class="va">priorInfo</span><span class="op">[</span>, <span class="st">"GENUS"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="set-up-normalization-and-da-methods">Set up normalization and DA methods<a class="anchor" aria-label="anchor" href="#set-up-normalization-and-da-methods"></a>
</h5>
<p>Both the normalization/scaling factors and the DA methods’
instructions are available since the dataset is the same used in the
previous section.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_normalizations</span></span>
<span><span class="co">#&gt; $norm_edgeR</span></span>
<span><span class="co">#&gt; $norm_edgeR$fun</span></span>
<span><span class="co">#&gt; [1] "norm_edgeR"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $norm_edgeR$method</span></span>
<span><span class="co">#&gt; [1] "TMM"</span></span>
<span><span class="va">my_methods</span></span>
<span><span class="co">#&gt; $DA_limma.1</span></span>
<span><span class="co">#&gt; $DA_limma.1$method</span></span>
<span><span class="co">#&gt; [1] "DA_limma"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_limma.1$assay_name</span></span>
<span><span class="co">#&gt; [1] "counts"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_limma.1$pseudo_count</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_limma.1$design</span></span>
<span><span class="co">#&gt; [1] "~1 + RSID + HMP_BODY_SUBSITE"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_limma.1$coef</span></span>
<span><span class="co">#&gt; [1] "HMP_BODY_SUBSITESupragingival Plaque"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_limma.1$norm</span></span>
<span><span class="co">#&gt; [1] "TMM"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_limma.1$weights</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_ALDEx2.1</span></span>
<span><span class="co">#&gt; $DA_ALDEx2.1$method</span></span>
<span><span class="co">#&gt; [1] "DA_ALDEx2"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_ALDEx2.1$assay_name</span></span>
<span><span class="co">#&gt; [1] "counts"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_ALDEx2.1$pseudo_count</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_ALDEx2.1$design</span></span>
<span><span class="co">#&gt; [1] "HMP_BODY_SUBSITE"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_ALDEx2.1$mc.samples</span></span>
<span><span class="co">#&gt; [1] 128</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_ALDEx2.1$test</span></span>
<span><span class="co">#&gt; [1] "wilcox"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_ALDEx2.1$paired.test</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_ALDEx2.1$denom</span></span>
<span><span class="co">#&gt; [1] "all"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DA_ALDEx2.1$contrast</span></span>
<span><span class="co">#&gt; [1] "HMP_BODY_SUBSITE"     "Supragingival Plaque" "Subgingival Plaque"  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $customMethod.1</span></span>
<span><span class="co">#&gt; $customMethod.1$method</span></span>
<span><span class="co">#&gt; [1] "DA_custom"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $customMethod.1$contrast</span></span>
<span><span class="co">#&gt; [1] "HMP_BODY_SUBSITE"     "Supragingival Plaque" "Subgingival Plaque"  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $customMethod.1$paired</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>To add the normalization/scaling factors to the object:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps_plaque_16S</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/runNormalizations.html" class="external-link">runNormalizations</a></span><span class="op">(</span></span>
<span>    normalization_list <span class="op">=</span> <span class="va">my_normalizations</span>, </span>
<span>    object <span class="op">=</span> <span class="va">ps_plaque_16S</span><span class="op">)</span></span></code></pre></div>
<p>A simple filter to remove rare taxa:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps_plaque_16S</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/filter_taxa.html" class="external-link">filter_taxa</a></span><span class="op">(</span>physeq <span class="op">=</span> <span class="va">ps_plaque_16S</span>, </span>
<span>    flist <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">3</span>, prune <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ps_plaque_16S</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 58 taxa and 60 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 60 samples by 8 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 58 taxa by 6 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; phy_tree()    Phylogenetic Tree: [ 58 tips and 57 internal nodes ]</span></span></code></pre></div>
<p>Perform DA analysis:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Plaque_16S_DA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/runDA.html" class="external-link">runDA</a></span><span class="op">(</span></span>
<span>    method_list <span class="op">=</span> <span class="va">my_methods</span>, </span>
<span>    object <span class="op">=</span> <span class="va">ps_plaque_16S</span>, </span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="testing-the-enrichment">Testing the enrichment<a class="anchor" aria-label="anchor" href="#testing-the-enrichment"></a>
</h5>
<p>We use the <code>createEnrichment</code> with the direction
parameter:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/createEnrichment.html" class="external-link">createEnrichment</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">Plaque_16S_DA</span>, <span class="co"># DA results</span></span>
<span>    priorKnowledge <span class="op">=</span> <span class="va">priorInfo</span>, <span class="co"># Pior knowledge</span></span>
<span>    enrichmentCol <span class="op">=</span> <span class="st">"Type"</span>, <span class="co"># name of the interesting column </span></span>
<span>    namesCol <span class="op">=</span> <span class="st">"newNames"</span>, <span class="co"># name of the new taxa names (opt)</span></span>
<span>    slot <span class="op">=</span> <span class="st">"pValMat"</span>, <span class="co"># Which output slot for DA detection</span></span>
<span>    colName <span class="op">=</span> <span class="st">"adjP"</span>, <span class="co"># Which column of the output slot</span></span>
<span>    type <span class="op">=</span> <span class="st">"pvalue"</span>, <span class="co"># What kind of measure it is</span></span>
<span>    <span class="co"># Optional, gives the sign of DA</span></span>
<span>    direction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="co"># name of the colums in statInfo matrix</span></span>
<span>        <span class="st">"logFC"</span>, <span class="co"># limma </span></span>
<span>        <span class="st">"effect"</span>, <span class="co"># ALDEx2</span></span>
<span>        <span class="st">"logFC"</span><span class="op">)</span>, <span class="co"># custom</span></span>
<span>    threshold_pvalue <span class="op">=</span> <span class="fl">0.1</span>, <span class="co"># adjP threshold for DA detection</span></span>
<span>    threshold_logfc <span class="op">=</span> <span class="fl">0</span>, <span class="co"># logFC threshold for DA detection</span></span>
<span>    top <span class="op">=</span> <span class="cn">NULL</span>, <span class="co"># select only the top features (opt)</span></span>
<span>    alternative <span class="op">=</span> <span class="st">"greater"</span>, <span class="co"># alternative test enrichment</span></span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Method: limma.TMM</span></span>
<span><span class="co">#&gt;  * 'adjP' column, as pvalue type, of pValMat matrix.</span></span>
<span><span class="co">#&gt;  * 'logFC' column, as direction, of statInfo matrix.</span></span>
<span><span class="co">#&gt;  * DA: , adjP&lt;=0.1, |logFC|&gt;=0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Method: ALDEx2.all.wilcox.paired</span></span>
<span><span class="co">#&gt;  * 'adjP' column, as pvalue type, of pValMat matrix.</span></span>
<span><span class="co">#&gt;  * 'effect' column, as direction, of statInfo matrix.</span></span>
<span><span class="co">#&gt;  * DA: , adjP&lt;=0.1, |effect|&gt;=0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Method: custom.paired</span></span>
<span><span class="co">#&gt;  * 'adjP' column, as pvalue type, of pValMat matrix.</span></span>
<span><span class="co">#&gt;  * 'logFC' column, as direction, of statInfo matrix.</span></span>
<span><span class="co">#&gt;  * DA: , adjP&lt;=0.1, |logFC|&gt;=0</span></span></code></pre></div>
<hr>
<blockquote>
<p><strong>More about the enrichment</strong></p>
<p>The produced <em>enrichment</em> object consists in a list of
elements as long as the number of tested methods:</p>
<ul>
<li><p>the <em>data</em> slot contains information for each feature.
P-values, adjusted p-values (or other statistics) in <em>stats</em>
column, log fold changes (or other statistics, if specified) in
<em>direction</em> column, differential abundance information in the
<em>DA</em> column (according to the thresholds), the variable of
interest for the enrichment analysis, and the name of the feature in the
<em>feature</em> column;</p></li>
<li><p>in the <em>tables</em> slot a maximum of <em>2 x (levels of
enrichment variable)</em> contingency tables (<em>2x2</em>) are
present;</p></li>
<li><p>in the <em>tests</em> slot, the list of Fisher exact tests
produced by the <code>fisher.test</code> function are saved for each
contingency table;</p></li>
<li><p>in the <em>summaries</em> slot, the first elements of the
contingency tables and the respective p-values are collected for
graphical purposes.</p></li>
</ul>
<p>Considering one of the methods, <code>limma.TMM</code>, we obtain 8
contingency tables. Both UP Abundant and DOWN Abundant taxa are found
and the enrichment variable has Aerobic, Anaerobic, F_Anaerobic, and
Unknown levels. For each level, we can build 2 contingency tables: one
for DOWN Abundant vs non-DOWN Abundant features and one for UP Abundant
vs non-UP Abundant features. The enrichment is tested using Fisher exact
test. The <code>plotContingency</code> function summarize all these
informations:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotContingency.html" class="external-link">plotContingency</a></span><span class="op">(</span>enrichment <span class="op">=</span> <span class="va">enrichment</span>, </span>
<span>    levels_to_plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Aerobic"</span>, <span class="st">"Anaerobic"</span><span class="op">)</span>, </span>
<span>    method <span class="op">=</span> <span class="st">"limma.TMM"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using .id, direction, level as id variables</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/plotContingency-1.png" alt="Contingency tables plot. Contingency tables for Aerobic and Anaerobic taxa found as differentially abundant by DESeq2.poscounts DA method. Fisher exact test has been performed on each contingency table. If the enrichment is signficantly present, the corresponding cell will be highlighted." width="700"><p class="caption">
Contingency tables plot. Contingency tables for Aerobic and Anaerobic
taxa found as differentially abundant by DESeq2.poscounts DA method.
Fisher exact test has been performed on each contingency table. If the
enrichment is signficantly present, the corresponding cell will be
highlighted.
</p>
</div>
</blockquote>
<hr>
<p>Using the <code>plotEnrichment</code> function we can see the
enrichment analysis results:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotEnrichment.html" class="external-link">plotEnrichment</a></span><span class="op">(</span>enrichment <span class="op">=</span> <span class="va">enrichment</span>, enrichmentCol <span class="op">=</span> <span class="st">"Type"</span>, </span>
<span>    levels_to_plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Aerobic"</span>, <span class="st">"Anaerobic"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/plotEnrichment-1.png" alt="Enrichment plot. Number of differentially abundant features, colored by aerobic or anaerobic metabolism, and directed according to differential abundance direction (UP or DOWN abundant)." width="480"><p class="caption">
Enrichment plot. Number of differentially abundant features, colored by
aerobic or anaerobic metabolism, and directed according to differential
abundance direction (UP or DOWN abundant).
</p>
</div>
<p>Since “Subgingival Plaque” is the reference level for each method,
the coefficients extracted from the methods are referred to the
“Supragingival Plaque” class. <code>limma.TMM</code> and
<code>ALDEx2.all.wilcox.paired</code> find, as expected, a statistically
significant (<span class="math inline">\(0.01 &lt; p \le 0.1\)</span>)
amount of DOWN Abundant Anaerobic features and UP Abundant Aerobic
features in Supragingival Plaque.</p>
<p>To investigate the DA features, the <code>plotMutualFindings</code>
function can be used to extract only features which are mutually found
as DA by more than 1 method:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotMutualFindings.html" class="external-link">plotMutualFindings</a></span><span class="op">(</span><span class="va">enrichment</span>, enrichmentCol <span class="op">=</span> <span class="st">"Type"</span>, </span>
<span>    levels_to_plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Aerobic"</span>, <span class="st">"Anaerobic"</span><span class="op">)</span>, n_methods <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/plotMutualFindings-1.png" alt="Mutual Findings plot. Number of differentially abundant features mutually found by 1 or more methods, colored by the differential abundance direction and separated by aerobic and anaerobic metabolism." width="576"><p class="caption">
Mutual Findings plot. Number of differentially abundant features
mutually found by 1 or more methods, colored by the differential
abundance direction and separated by aerobic and anaerobic metabolism.
</p>
</div>
<p>In this example, many Anaerobic genera and 3 Aerobic genera are found
as DA by more than 1 method simultaneously. Among them
<em>Prevotella</em>, <em>Treponema</em>, <em>Fusobacterium</em>,
<em>Catonella</em>, and <em>Dialister</em> genera are found as DOWN
Abundant in Supragingival Plaque by all methods. While the
<em>Actinomyces</em> genus is considered UP Abundant in Supragingival
Plaque.</p>
<hr>
<blockquote>
<p><strong>True and False Positives</strong></p>
<p>To evaluate the overall performances we can also study a statistic
based on the difference between putative True Positives (TP) and the
putative False Positives (FP). To build the matrix to plot, the
<code>createPositives</code> can be used. In details, the correctness of
the DA features is evaluated comparing the direction of the top ranked
features to the expected direction supplied by the user in the
<em>TP</em> and <em>FP</em> lists. The procedure is performed for
several thresholds of <em>top</em> parameter in order to observe a
trend, if present.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">positives</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/createPositives.html" class="external-link">createPositives</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">Plaque_16S_DA</span>, </span>
<span>    priorKnowledge <span class="op">=</span> <span class="va">priorInfo</span>, </span>
<span>    enrichmentCol <span class="op">=</span> <span class="st">"Type"</span>,</span>
<span>    namesCol <span class="op">=</span> <span class="st">"newNames"</span>, </span>
<span>    slot <span class="op">=</span> <span class="st">"pValMat"</span>, </span>
<span>    colName <span class="op">=</span> <span class="st">"adjP"</span>, </span>
<span>    type <span class="op">=</span> <span class="st">"pvalue"</span>,</span>
<span>    direction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"logFC"</span>, <span class="co"># limma</span></span>
<span>        <span class="st">"effect"</span>, <span class="co"># ALDEx2</span></span>
<span>        <span class="st">"logFC"</span><span class="op">)</span>, <span class="co"># custom</span></span>
<span>    threshold_pvalue <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    threshold_logfc <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    top <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq.int</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0</span>, to <span class="op">=</span> <span class="fl">20</span>, by <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>    alternative <span class="op">=</span> <span class="st">"greater"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    TP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DOWN Abundant"</span>, <span class="st">"Anaerobic"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UP Abundant"</span>, <span class="st">"Aerobic"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    FP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DOWN Abundant"</span>, <span class="st">"Aerobic"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UP Abundant"</span>, <span class="st">"Anaerobic"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Finally, the <code>plotPositives</code> function can be used to
summarize the methods’ performances. Higher values usually represents
better performances.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/benchdamic/man/plotPositives.html" class="external-link">plotPositives</a></span><span class="op">(</span><span class="va">positives</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="introduction_to_benchdamic_files/figure-html/plotPositives-1.png" alt="TP, FP differences plot. Differences between the number of True Positives and False Positives for several thresholds of the top ranked raw p-values (the top 2 lowest p-values, the top 4, 6, ..., 20) for each method." width="700"><p class="caption">
TP, FP differences plot. Differences between the number of True
Positives and False Positives for several thresholds of the top ranked
raw p-values (the top 2 lowest p-values, the top 4, 6, …, 20) for each
method.
</p>
</div>
<p>Under the working assumption (Aerobic Up Abundant in Supragingival
and Anaerobic Down Abundant in Supraginval)
<code>ALDEx2.all.wilcox.paired</code> confirms itself as a conservative
method. <code>limma.TMM</code>, along with the <code>custom</code>
method, are able to reach the highest values of the statistic locating
themselves on the upper part of the figure. This means that their
findings are in line with the <em>a priori</em> knowledge supplied by
the user.</p>
</blockquote>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="conclusions">Conclusions<a class="anchor" aria-label="anchor" href="#conclusions"></a>
</h4>
<p>In our dataset:</p>
<ul>
<li>
<p><em>NB</em>, <em>ZINB</em>, and <em>DM</em> models seem to have a
good fit when used for estimating the average counts and the proportion
of zero;</p>
<ul>
<li>Methods based on those parametric distributions should work
well.</li>
</ul>
</li>
<li>
<p><code>ALDEx2.all.wilcox.unpaired</code> had a conservative
behaviour when evaluating the proportion of False Positives;</p>
<ul>
<li>If you want to be sure about your findings, using a conservative
method could be a good choice.</li>
</ul>
</li>
<li><p><code>limma.TMM</code> results were concordant with those of the
<code>custom</code> method. <code>ALDEx2.all.wilcox.paired</code> had
the highest WMC values.</p></li>
<li>
<p>both <code>limma.TMM</code> and
<code>ALDEx2.all.wilcox.paired</code> found an enriched amount of Up
Abundant Aerobic taxa and Down Abundant Anaerobic taxa in Supragingival
Plaque;</p>
<ul>
<li><p>We can trust the findings of both methods;</p></li>
<li><p>We can even trust more the mutual findings;</p></li>
<li><p>If the assumptions are correct. ;)</p></li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:mcalgaro93@gmail.com" class="email">mcalgaro93@gmail.com</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>University of Verona, Department of Biotechnology<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Matteo Calgaro.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
